core/worker.go

```diff
@@ -1092,45 +1121,256 @@ func (w *worker) commitTransaction(env *environment, parent *types.WorkObject, t
+					return nil, false, fmt.Errorf("could not add new lock: %w", err)
+				}
+				// Store the new lockup key every time
+				env.coinbaseLockupsCreated[string(newCoinbaseLockupKey)] = *newCoinbaseLockupHash
+
+				if oldCoinbaseLockupHash != nil {
+					// We deleted (updated) the old lockup, write it to deleted list but only the first time
+					if _, exists := env.coinbaseLockupsDeleted[string(oldCoinbaseLockupKey)]; !exists {
+						if _, exists := env.coinbaseRotatedEpochs[string(newCoinbaseLockupKey)]; !exists {
+							env.coinbaseLockupsDeleted[string(oldCoinbaseLockupKey)] = *oldCoinbaseLockupHash
+							env.utxosDelete = append(env.utxosDelete, *oldCoinbaseLockupHash)
+						}
 					}
-					// the ETX hash is guaranteed to be unique
-					utxoHash := types.UTXOHash(tx.Hash(), outputIndex, types.NewUtxoEntry(types.NewTxOut(uint8(denomination), tx.To().Bytes(), lockup)))
-					env.utxosCreate = append(env.utxosCreate, utxoHash)
-					outputIndex++
+				} else {
+					// If we did not delete, we are rotating the epoch and need to store it
+					env.coinbaseRotatedEpochs[string(newCoinbaseLockupKey)] = struct{}{}
 				}
+				receipt := &types.Receipt{Type: tx.Type(), Status: types.ReceiptStatusLocked, GasUsed: gasUsedForCoinbase, TxHash: tx.Hash()} // todo: consider adding the reward to the receipt in a log
+
+				gasUsed := env.wo.GasUsed()
+				gasUsed += gasUsedForCoinbase
+				env.wo.Header().SetGasUsed(gasUsed)
+				env.gasUsedAfterTransaction = append(env.gasUsedAfterTransaction, gasUsed)
+				env.txs = append(env.txs, tx)
+				env.receipts = append(env.receipts, receipt)
+				return receipt.Logs, true, nil
+			} else {
+				// Coinbase data is either too long or too small
+				// Coinbase reward is lost
+				receipt := &types.Receipt{Type: tx.Type(), Status: types.ReceiptStatusFailed, GasUsed: gasUsedForCoinbase, TxHash: tx.Hash()}
+				gasUsed := env.wo.GasUsed()
+				gasUsed += gasUsedForCoinbase
+				env.wo.Header().SetGasUsed(gasUsed)
+				env.gasUsedAfterTransaction = append(env.gasUsedAfterTransaction, gasUsed)
+				env.txs = append(env.txs, tx)
+				env.receipts = append(env.receipts, receipt)
+				return []*types.Log{}, false, nil
 			}
-		}
-		gasUsed := env.wo.GasUsed()
-		if parent.NumberU64(common.ZONE_CTX) >= params.TimeToStartTx {
-			gasUsed += params.TxGas
-		}
-		env.wo.Header().SetGasUsed(gasUsed)
-		env.gasUsedAfterTransaction = append(env.gasUsedAfterTransaction, gasUsed)
-		env.txs = append(env.txs, tx)
 
-		return []*types.Log{}, false, nil
+		} else if tx.To().IsInQuaiLedgerScope() { // Quai coinbase
+			_, err := tx.To().InternalAndQuaiAddress()
+			if err != nil {
+				return nil, false, fmt.Errorf("coinbase tx %x has invalid recipient: %w", tx.Hash(), err)
+			}
+			var receipt *types.Receipt
+			if len(tx.Data()) == 1 {
+				// Coinbase has no extra data
+				// Coinbase is valid
+				receipt = &types.Receipt{Type: tx.Type(), Status: types.ReceiptStatusLocked, GasUsed: 21000, TxHash: tx.Hash()}
```

Comment:
probably better if the GasUsed was gasUsedForCoinbase instead of 21000

Reply:
Fixed